### ISSUE TITLE
Trailpack event `before`

### ISSUE THREAD

**Issue by jaumard** [trailsjs/trails#109]
It can be nice to have the ability to listen an event `before` for trailpack lifecycle.
Currently we can run our methods after another but not before. In some cases it can be useful to be executed before (change config file dynamically, change files ect....) 

We can easily imagine this kind of configuration : 
```
lifecycle: {
  emit: ['webserver:http:ready'],
  before: {
    initialize: ['trailpack:router:initialized']
  },
  after: {
    configure: ['trailpack:orm:configured']
  }
}
```

What do you think ?

**Comment by tjwebb** [comment]
```
before: {
    initialize: ['trailpack:router:initialized']
  },
```

I think we've already talked about this on gitter. This is the opposite of how I usually think about loading modules. If I'm writing a module, my module should declare what other things it depends on.

I'd need to see a specific instance where this allows you to accomplish something you can't currently accomplish with the current system. Keep in mind there are multiple lifecycle stages, so this could be achieved already in most cases.

Say module `A` injects some configuration stuff that module `B` needs, but module `B` isn't aware of `A`'s existence. Module `A` should inject this configuration during the `configure` lifecycle stage, and Module `B` can read that configuration during the `initialize` stage. `initialize` stage doesn't begin until after all trailpacks have completed their `configure` stage.

**Comment by jaumard** [comment]
Yes we did sorry but I can't find our talk so I decide to put it here. In fact it's maybe because I don't understand this like you do.
For me Trailpacks have 3 steps to do : 

1) Validation : config ect... 
2) Configure : prepare some stuff before init 
3) Initlize : really launch the Trailpack to start doing what is suppose to (but initialize do only the start and don't do stuff) 

With this in my mind it's difficult to me to understand because if a Trailpack really configure itself in the `configure` method other Trailpack have to do the job before this method and not before the `initialize` method.
But in all current Trailpack all stuff is made under `initialize` so yeah I can do my stuff in the `configure`.

If I take the example of hapi (or express) all stuff is in the `initialize` but naturally when I code it the `initialize` will only start the server and all other things will be in the `configure` to prepare the `initialize`(create server, register views, middleware...).

So maybe it's just wrong lifecycle methods name or need more documentation about what need to do these methods. I find my proposition more "natural" than the actual (but only my opinion ^^)

Thanks for the explanation btw ! It's more clear now

**Comment by jaumard** [comment]
Maybe this can resolved by my proposal for Trailpack hooks here : https://github.com/trailsjs/trails/issues/129#issuecomment-208194308
